File: WasherCheck/src/main/java/net/zdremann/util/AsyncTaskResult.java
package net.zdremann.util;

public final class AsyncTaskResult<T> {
    private final T result;
    private final Exception error;
    private final boolean isResult;

    public boolean isResult() { return isResult; }

    public T getResult() {
        return result;
    }

    public Exception getError() {
        return error;
    }

    public AsyncTaskResult(T result) {
        this.isResult = true;
        this.error = null;
        this.result = result;
    }

    public AsyncTaskResult(Exception error) {
        this.isResult = false;
        this.error = error;
        this.result = null;
    }
}

File: WasherCheck/src/main/kotlin/net/zdremann/util/AsyncTaskResult.kt
package net.zdremann.util

class AsyncTaskResult<out T> private constructor(val value: T?, val error: Exception?) {
    companion object {
        @JvmStatic
        fun <T> value(value: T): AsyncTaskResult<T> = AsyncTaskResult(value, null)

        @JvmStatic
        fun <T> error(error: Exception): AsyncTaskResult<T> = AsyncTaskResult(null, error)
    }

    val isValue: Boolean
        get() = error == null
}
